app/src/main/java/com/hedvig/app/feature/offer/usecase/GetPostSignDependenciesUseCase.kt

```diff
@@ -1,35 +1,30 @@
 package com.hedvig.app.feature.offer.usecase
 
+import arrow.core.Either
 import arrow.core.NonEmptyList
+import arrow.core.left
 import com.hedvig.app.feature.offer.OfferRepository
+import com.hedvig.app.feature.offer.model.OfferModel
 import com.hedvig.app.feature.offer.model.quotebundle.PostSignScreen
-import kotlinx.coroutines.flow.first
+import com.hedvig.app.util.ErrorMessage
+import kotlinx.coroutines.flow.firstOrNull
 
 class GetPostSignDependenciesUseCase(
     private val offerRepository: OfferRepository,
 ) {
-    sealed class Result {
-        data class Success(
-            val postSignScreen: PostSignScreen,
-            val displayName: String,
-        ) : Result()
+    data class Result(
+        val postSignScreen: PostSignScreen,
+        val displayName: String,
+    )
 
-        object Error : Result()
-    }
+    suspend operator fun invoke(quoteIds: List<String>): Either<ErrorMessage, Result> = offerRepository
+        .offer(NonEmptyList.fromListUnsafe(quoteIds))
+        .firstOrNull()
```

Comment:
Hmm, I am pretty sure in this case the firstOrNull() will never be null since if it would be null we'd be getting a runtime crash since fromListUnsafe throws when the list passed into it is empty.

I think it's be best if we either make the quoteIds also a NEL and use the safe alternatives when creating it. Or if we want to keep the function signature the same, we could replace this with something safe like:
```kotlin
suspend operator fun invoke(quoteIds: List<String>): Either<ErrorMessage, Result> {
    if (quoteIds.isEmpty()) {
        return ErrorMessage().left()
    }
    return offerRepository
        .offer(NonEmptyList.fromListUnsafe(quoteIds))
        .firstOrNull()
        ?.map { it.toResult() }
        ?: ErrorMessage().left()
}
```
Like you did in some other place in the code.

After the answer in Arrow slack I can see this `either{}` exists too, which could make something like this possible as well:

```kotlin
suspend operator fun invoke(quoteIds: List<String>): Either<ErrorMessage, Result> {
    return either {
        val ids = NonEmptyList.fromList(quoteIds).toEither { ErrorMessage() }.bind()
        val result = offerRepository
            .offer(ids)
            .firstOrNull()
            ?.map { it.toResult() }
            ?: ErrorMessage().left()
        result.bind()
    }
}
```

Interesting, but probably not for this use case. But interesting nonetheless.

Reply:
Hey @StylianosGakis,
Another small suggestion here is that you can use `ensureNotNull`.
This eliminates the need for `map`, and lifting `ErrorMessage()` into `left()`.

```kotlin
suspend operator fun invoke(quoteIds: List<String>): Either<ErrorMessage, Result> = either {
        val ids = NonEmptyList.fromList(quoteIds).toEither { ErrorMessage() }.bind()
        val result = offerRepository
            .offer(ids)
            .firstOrNull()
            ?.bind()?.toResult()
         ensureNotNull(result) { ErrorMessage() }
}
```